### PR TITLE
feat: add ROS 2 service introspection for Service and Client

### PR DESCRIPTION
**Pull request by jbcpollak**
## Summary

Brings ROS 2 Service Introspection (Humble+) to `rclgo`. Services and clients can now publish service-event messages on the hidden `<name>/_service_event` topic, so requests/responses can be monitored with tools like `ros2 service echo`.

Wraps `rcl_service_configure_service_introspection` / `rcl_client_configure_service_introspection` from `rcl/service.h` and `rcl/client.h`.

## API

```go
type ServiceIntrospectionState int // mirrors rcl_service_introspection_state_t
const (
    ServiceIntrospectionOff      // disabled
    ServiceIntrospectionMetadata // metadata only
    ServiceIntrospectionContents // metadata + request/response contents
)

// Declarative: set initial state at construction
ServiceOptions{ Introspection: ServiceIntrospectionContents }
ClientOptions{  Introspection: ServiceIntrospectionMetadata }

// Imperative: reconfigure at runtime
func (s *Service) ConfigureIntrospection(state ServiceIntrospectionState) error
func (s *Service) IntrospectionState() ServiceIntrospectionState
func (c *Client)  ConfigureIntrospection(state ServiceIntrospectionState) error
func (c *Client)  IntrospectionState() ServiceIntrospectionState
```

## Implementation notes

- **CGO**: `rcl/service.h` and `rcl/client.h` are already in the cgo preamble and transitively include `service_introspection.h`, so no new includes were needed. `publisher_options` is passed by value (no allocation to free); the type-support pointer reuses the existing cast idiom.
- **Stored state**: `Service` and `Client` now keep the `types.ServiceTypeSupport` required by the rcl calls; event timestamps use the node context clock (same source as `ActionServer`).
- **Concurrency**: the rcl call is documented not thread-safe, so each struct serializes `ConfigureIntrospection` with a mutex and caches the current state. Calls after `Close()` return the standard closed-sentinel error.
- **Lifecycle**: the existing `rcl_*_fini` calls tear down the introspection publisher — no extra cleanup needed.

## Testing

- `go build ./pkg/rclgo/` passes; `gofmt` clean.
- Added `TestServiceIntrospection` covering default-off, runtime transitions, and enabling via options.
- Note: the full `go test ./...` suite requires the `test_msgs` ROS package, which was not installed in the dev container used here.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/rclgo/rcl.go [hunk 1]
```diff
@@ -924,10 +975,58 @@ func (n *Node) NewService(
 	if retCode != C.RCL_RET_OK {
 		return nil, errorsCastC(retCode, "failed to create service")
 	}
+	if options.Introspection != ServiceIntrospectionOff {
+		if err = s.ConfigureIntrospection(options.Introspection); err != nil {
+			return nil, err
```
Comment (fastcat) [inline pkg/rclgo/rcl.go:981]: Need to de-init the service in this error path, else it's a leak

File: pkg/rclgo/rcl.go [hunk 2]
```diff
@@ -1038,10 +1154,58 @@ func (n *Node) NewClient(
 	if rc != C.RCL_RET_OK {
 		return nil, errorsCastC(rc, "failed to create client")
 	}
+	if options.Introspection != ServiceIntrospectionOff {
+		if err = c.ConfigureIntrospection(options.Introspection); err != nil {
+			return nil, err
```
Comment (fastcat) [inline pkg/rclgo/rcl.go:1162]: Same deinit issue here

File: pkg/rclgo/service_test.go
Comment (fastcat) [inline pkg/rclgo/service_test.go:1]: Just because I'm paranoid, should add a test that validates that the zero value of `ServiceIntrospectionState` is `ServiceIntrospectionOff`. I would be shocked if the RCL enum somehow had off not be zero, but ROS does weird stuff sometimes...

**Review by fastcat — CHANGES_REQUESTED** [review]

File: pkg/rclgo/rcl.go (hunk 1 above)
Comment (jbcpollak) [reply pkg/rclgo/rcl.go:981]: @fastcat doesn't the `defer onErr(&err, s.Close)` on line 914 (pre) or 965 (post commit) handle this? Likewise there's the same line for the NewClient().

Comment (fastcat) [reply pkg/rclgo/rcl.go:981]: Yeah, it does.

I hate that anti-idiomatic pattern, but it's the norm in this repo 🤷‍♂️

**Review by fastcat — APPROVED** [review]
LGTM, assuming we've tested that it works when integrated into the stack :)

File: pkg/rclgo/rcl.go (hunk 1 above)
Comment (jbcpollak) [reply pkg/rclgo/rcl.go:981]: what would the idomatic fix be? I thought defers were the way to go here? we could have claude do a sweep to make it idiomatic if you want

Comment (fastcat) [reply pkg/rclgo/rcl.go:981]: I don't think we should do anything here right now. Maybe at most put a comment pointing at the defer.

Idiomatic would be to be explicit and put the error handling explicitly before the return.

**Comment by jbcpollak** [comment]
I asked Claude for more rigorous testing and this is what it said. I think the payload-level testing would be a good followup:

I dug into feasibility, and my recommendation is now more nuanced than "yes, add more tests." Here's the honest picture:

The existing tests already cover more than state-caching. ConfigureIntrospection returning nil is meaningful: rcl_service_configure_service_introspection creates (or for Off, destroys) the hidden introspection publisher internally and returns an error if that fails. So the current tests — transitions across Off/Metadata/Contents, and enabling via options at construction — already exercise the real CGO path and confirm rcl actually (de)allocates the publisher for every state. That's genuine behavioral coverage.

The only untested aspect is observable event flow, and both ways to test it have problems:
- Payload-level (assert real ServiceEvent contents): not feasible — the _Event / service_msgs Go bindings aren't generated.
- Graph-level (assert /<service>/_service_event appears in GetTopicNamesAndTypes): feasible in principle, but it's a discovery-timing test (needs polling, inherently flaky) and depends on rmw hidden-topic mangling semantics that differ per middleware — and we just switched this suite to Zenoh and stabilized CI. Real risk of reintroducing flakiness for modest added assurance.

My recommendation: the current coverage is reasonable to ship this PR. I would not rush in a graph-discovery test right after stabilizing CI. Better as a follow-up: implement the payload-level test properly once _Event bindings are generated — that's the test that actually proves events flow, and it's deterministic (you receive the message or you don't).

So: nothing missing that blocks the PR, in my view. If you'd still like the end-to-end signal now, I'd add the graph-level test with a generous poll-timeout and tolerant name matching — your call, given the flakiness trade-off.

**Comment by fastcat** [comment]
> rcl_service_configure_service_introspection creates (or for Off, destroys) the hidden introspection publisher internally and returns an error if that fails.

Does this mean we need to wire a call to configure it to Off when we destroy the service server/client? Or are the RCL internals already doing that for us when the Close (fini?) call goes throught?

**Comment by SimNine** [comment]
> > rcl_service_configure_service_introspection creates (or for Off, destroys) the hidden introspection publisher internally and returns an error if that fails.
> 
> Does this mean we need to wire a call to configure it to Off when we destroy the service server/client? Or are the RCL internals already doing that for us when the Close (fini?) call goes throught?

Looks like the hidden publisher gets destroyed on deallocation

```
rcl_ret_t
rcl_client_fini(rcl_client_t * client, rcl_node_t * node)
{

  ...stuff...

  if (client->impl) {
    rcl_allocator_t allocator = client->impl->options.allocator;
    rmw_node_t * rmw_node = rcl_node_get_rmw_handle(node);
    if (!rmw_node) {
      return RCL_RET_INVALID_ARGUMENT;
    }

    rcl_ret_t rcl_ret = unconfigure_service_introspection(node, client->impl, &allocator);
    if (RCL_RET_OK != rcl_ret) {
      RCL_SET_ERROR_MSG(rcl_get_error_string().str);
      result = rcl_ret;
    }

   ...more stuff...
```

**Review by SimNine — APPROVED** [review]
